### PR TITLE
Use Bolt instead of LevelDB

### PR DESCRIPTION
**Pull request by calmh**
So, I'm putting this one up here for people to see. Diff is huge as this is essentially a full line by line rewrite of leveldb.go (our db driver) into boltdb.go plus a bunch of assorted changes around it. I guess the advantages I see with bolt are that it uses less CPU (around 30% faster in the TestBenchmarkTransferManyFiles test on my box), a bit less memory I think and uses a single file instead of a directory. API is a bit nicer too for the most part, although there are things around simultaneous read+write access that can be tricky. Disadvantage is that it currently doesn't build on Solaris, although that should be a fairly simple fix soon...

### CODE REVIEW

File: cmd/syncthing/main.go [hunk 1]
```diff
@@ -352,8 +353,8 @@ func main() {
 		l.Infof("Upgrade available (current %q < latest %q)", Version, rel.Tag)
 
 		if doUpgrade {
-			// Use leveldb database locks to protect against concurrent upgrades
-			_, err = leveldb.OpenFile(locations[locDatabase], &opt.Options{OpenFilesCacheCapacity: 100})
+			// Use database locks to protect against concurrent upgrades
+			ldb, err := db.NewBoltDB(locations[locDatabase])
```
Comment (AudriusButkevicius) [inline cmd/syncthing/main.go]: bdb or db?

File: cmd/syncthing/main.go [hunk 2]
```diff
@@ -638,6 +631,15 @@ func syncthingMain() {
...
 
+	for _, folderCfg := range cfg.Folders() {
+		for _, device := range folderCfg.DeviceIDs() {
+			if device == myID {
+				continue
+			}
+			m.Index(device, folderCfg.ID, nil, 0, nil)
+		}
+	}
+
```
Comment (AudriusButkevicius) [inline cmd/syncthing/main.go]: We should do this before starting the model, just in case.

File: internal/db/blockmap_types.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// Copyright (C) 2015 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package db
```
Comment (AudriusButkevicius) [inline internal/db/blockmap_types.go]: Needs gen headers

File: internal/db/blockmap_test.go [hunk 4]
```diff
@@ -118,142 +106,125 @@ func TestBlockMapAddUpdateWipe(t *testing.T) {
...
+		m.Add([]protocol.FileInfo{f})
 	}
 
-	f1.Flags = 0
+	b.ReportAllocs()
 }
 
-func TestBlockFinderFix(t *testing.T) {
-	db, f := setup()
+/*
```
Comment (AudriusButkevicius) [inline internal/db/blockmap_test.go]: We have version control for this, might as well scrap the whole section.

File: internal/db/boltdb.go [hunk 5]
```diff
@@ -0,0 +1,837 @@
...
+	globalBucketID = []byte("global")
+	folderBucketID = []byte("folders")
+)
+
+func folderBucket(tx *bolt.Tx, folder []byte) *bolt.Bucket {
+	return tx.Bucket(folderBucketID).Bucket(folder)
+}
+
+func folderDeviceBucket(tx *bolt.Tx, folder, device []byte) *bolt.Bucket {
+	bkt, err := tx.Bucket(folderBucketID).Bucket(folder).CreateBucketIfNotExists(device)
```
Comment (AudriusButkevicius) [inline internal/db/boltdb.go]: folderBucket instead of repetition maybe?

**Comment by AudriusButkevicius** [comment]
So I am all four for reducing resource usage.
I wonder how does it perform on a low-powered system compared to leveldb.

Obviously the diff is too huge to be properly reviewed, hence if integration tests pass, I guess I am happy.

**Comment by plouj** [comment]
Would now be a good time to write an generic DB interface so that the higher level code (syncthing) doesn't have to depend on the low level DB implementation details (a la apply inversion of dependency)?

**Comment by AudriusButkevicius** [comment]
It's down to the constraints of the database, the reason there is so much boilerplate is because bolt doesn't allow write transactions while a read transaction is happening, something leveldb didn't care about.

**Comment by calmh** [comment]
Yeah this is a small step in that direction, as it creates an actual database object with methods. I'm not proposing this for merge yet (at the very least I want it to build on Solaris as I'm one of the 0.1% of users who run that), it needs a lot more testing and so on. But I'm keeping it alive as a card up our sleeves for the future.

**Comment by AudriusButkevicius** [comment]
I think we either merge this or we don't as otherwise it will just gather dust and never make it.

**Comment by dcarosone** [comment]
another solarish user here, FYI..  (SmartOS and OmniOS)

On 21 July 2015 at 07:08, Jakob Borg <notifications@github.com> wrote:

> Yeah this is a small step in that direction, as it creates an actual
> database object with methods. I'm not proposing this for merge yet (at the
> very least I want it to build on Solaris as I'm one of the 0.1% of users
> who run that), it needs a lot more testing and so on. But I'm keeping it
> alive as a card up our sleeves for the future.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/syncthing/syncthing/pull/2073#issuecomment-123039606>.
>



-- 
Dan.

File: internal/db/blockmap_test.go (hunk 4 above)
Comment (calmh) [inline internal/db/blockmap_test.go]: Hmm, I think this is actually something that I've not implemented yet, I think it's still necessary...

**Comment by calmh** [comment]
Well, I'm running it myself so I keep it up to date and working. I'll rectify the stuff above and then we can revisit for a v0.12, perhaps with the other major changes brewing (relay and so on).

**Comment by alex2108** [comment]
>  I wonder how does it perform on a low-powered system compared to leveldb.
I could test on a raspberry pi, every bit of less CPU usage is valuable there.
Rescanning everything will be a pain, but I think I have to go through this once, but it has to wait until night so I don't generate conflicts while it's scanning.

**Comment by alex2108** [comment]
First test on raspberry pi was done after more than 4h of initial scanning and also 1h or more of initial metadata sync (files were all the same), since the initial scan of the leveldb release is long ago I don't remember how long that needed, but it's no huge difference.

when first looking at memory I was a bit shocked, but the shared part can be used if it's needed by other processes, so it's around the same like leveldb.
```
  PID USER      PR  NI  VIRT  RES  SHR S  %CPU %MEM    TIME+  COMMAND           
22019 pi        39  19 1291m 221m 126m S   0,3 59,6  53:18.12 syncthing         
```

index exchange when connecting seems to be a bit better and also file transfer maybe around 10% faster (tested with one 100MB file).

Another thing that I noticed and i'm pretty sure I never saw that on the leveldb version: between the 100% CPU usage of syncthing there are also some seconds where it's at almost 100% i/o wait during index exchange.

**Comment by AudriusButkevicius** [comment]
https://forum.syncthing.net/uploads/default/original/2X/9/97431bb03e95dfa07ce412f3c7b078386c5fa4e3.log

**Comment by calmh** [comment]
Panic look like things are just taking a very, very long time on that box and triggering the deadlock detector.

**Comment by calmh** [comment]
Not going to happen for now

**Pull request closed without merging**